### PR TITLE
Johnstill/rw 154 redux

### PR DESCRIPTION
**Pull request by johnstill**
This is a pretty extensive reworking of the cohort builder search app, so I'll give a quick, high level summary of the changes.

First, the +/- numbers are misleading; `package-lock.json` accounts for ~1500 additions so I think the actual +/- is pretty close to break even, maybe even a net reduction of a few hundred.

Second, all files I refer to are relative to `workbench/ui/src/app/cohort-search/`.

Documentation on the new technologies introduced and (briefly) why we're using them:

  * [Redux](http://redux.js.org/) - state management library / pattern
  * [angular-redux](https://github.com/angular-redux/store#getting-started) - functions and objects for using redux with Angular 2+
  * [redux-observable](https://redux-observable.js.org/) - provides a representation of redux's action stream as an observable, thus providing a method by which to attach handlers to asynchronous events in the action stream (plays a similar role as `redux-thunk` + promises)
  * [immutable.js](https://facebook.github.io/immutable-js/docs/#/) - redux requires state immutability; immutable.js provides it

The basic pattern is this: application state is modeled as a single tree like structure called a *store*.  Thus, any particular aspect of the state can be found by traversing that tree via a `keyPath` - there are many `keyPaths` in the code, they are understood by `immutable` and `angular-redux` in particular.  The application's initial state and some `keyPaths` are given in `store.ts`.  The transition from state to state is handled deterministically by a pure function called a *reducer* located in `reducer.ts` and a set of stream watchers called *epics* in `epics.ts` - the reducer handles synchronous and pure state transitions, the epics are designed to handle asynchronous actions (and are therefore the application's interface to the API).  The components interface with the store is through *action creators*, all of which are bundled into a service in `actions.ts`.

The general rule is that if a component needs state from the store, it asks for it via the stuff `angular-redux` provides, and if it needs to alter some state that is relevant to components *outside the component tree of which it is the root*, that state should go in the store and be altered by dispatching actions, which redux's machinery will hand off to the appropriate reducer / epic.  Local state can stay local.

This is a large refactor (larger than I had intended or wanted) of the existing functionality and there's a bit of new functionality, or at least the beginnings of it - error handling, for example.  I recommend cloning down a local copy of the branch and running it.  If you do so, I recommend you install the following Chrome browser extensions:

  * [Redux Devtools](https://chrome.google.com/webstore/detail/redux-devtools/lmhkpmbekcpmknklioeibfkpmmfibljd?hl=en) - allow you to inspect state and sniff the event stream
  * [Augury](https://chrome.google.com/webstore/detail/augury/elgalmkoelokbchhkhacckoklkejnhcd?hl=en) - Angular Devtools, you can get a useful overview of the component hierarchy here
  * [Immutable Object Formatter](https://chrome.google.com/webstore/detail/immutablejs-object-format/hgldghadipiblonfkkicmgcbbijnpeog) - immutable.js objects are by default pretty hard to parse in the console, this helps

Be aware that about ~150k subjects in memory (not necessarily in the cohort) the redux devtools will probably crash (/shrug, /le-sigh) but the app itself should continue to run fine.

Questions & comments are welcome.

EDIT: I _really_ don't recommend looking at any individual commits; they're all pretty broken taken separately.

### CODE REVIEW

File: ui/src/app/app.module.ts [hunk 1]
```diff
@@ -22,7 +22,7 @@ import {environment} from 'environments/environment';
 
 /* Our Modules */
 import {AppRoutingModule} from 'app/app-routing.module';
-import {CohortSearchModule} from './cohort-search/cohort-search.module';
+import {CohortSearchModule} from './cohort-search/module';
```
Comment (freemabd1) [inline ui/src/app/app.module.ts]: Should module name match the folder name?  In this case cohort-search(./cohort-search/module) doesn't match naming convention of (./cohort-review/cohort-review.module). This is an angular convention to name components/modules/etc.. the same as folder name. I'm not attached to this by an means but i like consistency.

File: ui/src/app/cohort-search/add-criteria.component.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
+import {Component, Input} from '@angular/core';
```
Comment (freemabd1) [inline ui/src/app/cohort-search/add-criteria.component.ts]: Should this component have it's own directory? It's currently the only component without one.

File: ui/src/app/cohort-search/charts/charts.component.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+      <app-gender-chart [data]="genderCounts">
+      </app-gender-chart>
+    </div>
+    <div>&nbsp;</div>
+    <div class="box">
+      <app-race-chart [data]="raceCounts">
+      </app-race-chart>
+    </div>
+  `,
+  styles: [`.box { border: 1px solid #d7d7d7; }`],
```
Comment (freemabd1) [inline ui/src/app/cohort-search/charts/charts.component.ts]: I tend to prefer having html and css files instead of inlining them. It can get messy as the app grows and we have to keep changing the html and adding styles. In this case it may be fine since charts probably won't change that much.

File: ui/src/app/cohort-search/cohort-builder/cohort-builder.component.ts [hunk 4]
```diff
@@ -1,45 +1,47 @@
...
+import {
+  CohortSearchState,
+  inclusionGroups,
+  exclusionGroups,
+  wizardOpen,
+  subjects,
+} from '../store';
 
 @Component({
   selector: 'app-search',
```
Comment (freemabd1) [inline ui/src/app/cohort-search/cohort-builder/cohort-builder.component.ts]: Should we change this to match the component(app-cohort-builder)?

File: ui/src/app/cohort-search/cohort-builder/cohort-builder.component.ts [hunk 5]
```diff
@@ -49,57 +51,13 @@ export class CohortBuilderComponent implements OnInit, OnDestroy {
...
-
-  removeExclusionGroup(searchGroup: SearchGroup) {
-    const index: number = this.exclusionGroups.indexOf(searchGroup);
-    if (index !== -1) {
-      this.exclusionGroups.splice(index, 1);
-    }
-    this.changeDetectorRef.detectChanges();
-    this.updateTotalSet();
-    this.updateCharts();
+    this.subscription = this.state$.subscribe(state => {
```
Comment (freemabd1) [inline ui/src/app/cohort-search/cohort-builder/cohort-builder.component.ts]: How do subscriptions get unsubscribed with redux?

File: ui/src/app/cohort-search/criteria-tree/node-info.component.html [hunk 6]
```diff
@@ -0,0 +1,37 @@
+<div class="row clr-treenode-link">
```
Comment (freemabd1) [inline ui/src/app/cohort-search/criteria-tree/node-info.component.html]: Should we have directories for each component?

File: ui/src/app/cohort-search/criteria-tree/node.component.ts [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+import {Subscription} from 'rxjs/Subscription';
+
+import {CohortSearchActions} from '../actions';
+import {CohortSearchState} from '../store';
+
+import {Criteria} from 'generated';
+
+
+@Component({
+  selector: 'app-tree-node',
```
Comment (freemabd1) [inline ui/src/app/cohort-search/criteria-tree/node.component.ts]: Should the selector be app-node to match the node.component.ts name

File: ui/src/app/cohort-search/criteria-tree/root.component.ts [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+} from '@angular/core';
+import {NgRedux} from '@angular-redux/store';
+import {Subscription} from 'rxjs/Subscription';
+
+import {CohortSearchActions} from '../actions';
+import {CohortSearchState} from '../store';
+
+
+@Component({
+  selector: 'app-tree-root',
```
Comment (freemabd1) [inline ui/src/app/cohort-search/criteria-tree/root.component.ts]: selector: 'app-root'?

File: ui/src/app/cohort-search/epics.ts [hunk 9]
```diff
@@ -0,0 +1,53 @@
+import {ActionsObservable} from 'redux-observable';
```
Comment (freemabd1) [inline ui/src/app/cohort-search/epics.ts]: Should we have a directory for all the redux ts files? Like actions.ts, epics.ts, reducer.ts and store.ts

File: ui/src/app/cohort-search/wizard-select/wizard-select.component.ts [hunk 10]
```diff
@@ -1,39 +0,0 @@
...
-  readonly criteriaTypes = CRITERIA_TYPES;
-
-  @ViewChild('parent', {read: ViewContainerRef})
-  parent: ViewContainerRef;
-
-  constructor(private broadcastService: BroadcastService,
-              private componentFactoryResolver: ComponentFactoryResolver) { }
-
-  openWizard(criteriaType: string) {
-    const wizardModalComponent = this.componentFactoryResolver.resolveComponentFactory(
```
Comment (freemabd1) [inline ui/src/app/cohort-search/wizard-select/wizard-select.component.ts]: So we can talk about this in the morning, but just curious how this wizard component is getting dynamically created in the new redux code?

**Comment by osgsusan** [comment]
Great summary, your time spent is much appreciated.
